src/esp/gfx/CubeMap.h

```diff
@@ -0,0 +1,207 @@
+   * @return Reference to the cubemap texture
+   */
+  Magnum::GL::CubeMapTexture& getTexture(TextureType type);
+
+#ifndef MAGNUM_TARGET_WEBGL
+  /**
+   * ```
+   *           +----+
+   *           | -Y |
+   * +----+----+----+----+
+   * | -Z | -X | +Z | +X |
+   * +----+----+----+----+
+   *           | +Y |
+   *           +----+
+   * ```
+   * NOTE: +Y is top
+   * @brief save the cubemap texture based on the texture type
+   * @param type, texture type
+   * @param imageFilePrefix, the filename prefix
+   * The 6 image files then would be:
+   * {imageFilePrefix}.{texType}.+X.png
+   * {imageFilePrefix}.{texType}.-X.png
+   * {imageFilePrefix}.{texType}.+Y.png
+   * {imageFilePrefix}.{texType}.-Y.png
+   * {imageFilePrefix}.{texType}.+Z.png
+   * {imageFilePrefix}.{texType}.-Z.png
+   * @return true, if success, otherwise false
+   *
+   * NOTE:
+   * In this version, it will lose precision when depth texture is saved to the
+   * disk. This is because the pixel format for depth texture is R32F with 1 bit
+   * of sign, 23 bits of mantissa and 8 bits of exponent while the radiance HDR
+   * is 8 bits of mantissa and 8 bits of exponent. So when it is saved, only 8
+   * bits is saved to the R channel (GB channels just repeat the R channel
+   * twice).
+   */
+  // TODO: save HDR textures in EXR format
+  bool saveTexture(TextureType type, const std::string& imageFilePrefix);
+#endif
+
+  /**
+   * ```
+   *           +----+
+   *           | -Y |
+   * +----+----+----+----+
+   * | -Z | -X | +Z | +X |
+   * +----+----+----+----+
+   *           | +Y |
+   *           +----+
+   * ```
+   * NOTE: +Y is top
+   * @brief load cubemap texture from external images
+   * @param type can be "rgba", "depth", or "objectId" (TODO)
+   * @param imageFilePrefix, the prefix of the image filename
+   * @param imageFileExtension, the image filename extension (such as "png",
+   * "jpg")
+   * @return true if succeeded, otherwise false
+   * The 6 image files then would be:
+   * {imageFilePrefix}.{texType}.+X.{imageFileExtension}
+   * {imageFilePrefix}.{texType}.-X.{imageFileExtension}
+   * {imageFilePrefix}.{texType}.+Y.{imageFileExtension}
+   * {imageFilePrefix}.{texType}.-Y.{imageFileExtension}
+   * {imageFilePrefix}.{texType}.+Z.{imageFileExtension}
+   * {imageFilePrefix}.{texType}.-Z.{imageFileExtension}
+   */
+  void loadTexture(TextureType type,
+                   const std::string& imageFilePrefix,
+                   const std::string& imageFileExtension);
+
+  /**
+   * @brief Render to cubemap texture using the camera
+   * @param camera, a cubemap camera
+   * NOTE: It will NOT automatically generate the mipmap for the user
+   */
+  void renderToTexture(CubeMapCamera& camera,
+                       scene::SceneGraph& sceneGraph,
+                       RenderCamera::Flags flags);
+
+ private:
+  Flags flags_;
+  int imageSize_ = 0;
+  std::map<TextureType, std::unique_ptr<Magnum::GL::CubeMapTexture>> textures_;
```

Comment:
Not a fan of `std::map`, especially if you index it by a monotonic enum; the unique pointer is totally unnecessary also. Could be just a plain C array indexed by `TextureType`, if you want extra robustness you could wrap it in some getter, like

```cpp
enum class TextureType: uint8_t { 
    Color,
    Depth,
    ...
    Count
 };

Magnum::GL::CubeMapTexture textures_[uint8_t(TextureType::Count)];

Magnum::GL::CubeMapTexture& texture(TextureType type) {
    CORRADE_INTERNAL_ASSERT(uint8_t(type) < Containers::arraySize(textures_));
    return textures_[uint8_t(type)];
}
```